### PR TITLE
Inventory/Crafting: Supply Satchel support

### PR DESCRIPTION
**Pull request by kirmmin**
This PR enables the ability to use the Supply Satchel for item storage. You can move items between the Inventory and the Supply Satchel. There is no consumption of items in the Supply Satchel through Crafting at this time.

### CODE REVIEW

File: Source/NexusForever.WorldServer/Game/Entity/SupplySatchelManager.cs [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+        /// Builds the Network Packet for <see cref="ServerPlayerCreate"/>
+        /// </summary>
+        public ushort[] BuildNetworkPacket()
+        {
+            ushort[] tradeskillMaterialsCount = new ushort[512];
+
+            foreach ((ushort materialId, TradeskillMaterial material) in tradeskillMaterials)
+                tradeskillMaterialsCount[materialId] = material.Amount;
+
+            string text = "";
```
Comment (Rawaho) [inline Source/NexusForever.WorldServer/Game/Entity/SupplySatchelManager.cs]: Seems like debug logging that can be removed?

File: Source/NexusForever.WorldServer/Game/Entity/SupplySatchelManager.cs [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+        /// </summary>
+        /// <returns>Returns the remainder that could not be added.</returns>
+        public uint AddAmount(Item item, uint amount)
+        {
+            if (amount == 0)
+                throw new ArgumentException("amount must be more than 0");
+
+            TradeskillMaterialEntry entry = GameTableManager.Instance.TradeskillMaterial.Entries.SingleOrDefault(i => i.Item2IdStatRevolution == item.Entry.Id);
+            if (entry == null)
+                throw new ArgumentNullException("entry");
```
Comment (Rawaho) [inline Source/NexusForever.WorldServer/Game/Entity/SupplySatchelManager.cs]: ArgumentNullException refers to a parameter in the function.
You would use it like ArgumentNullException(nameof(parameterName))
In this case it's the invalid exception type to be used here

File: Source/NexusForever.WorldServer/Game/Entity/SupplySatchelManager.cs [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+        /// </summary>
+        /// <returns>Returns the remainder that could not be added.</returns>
+        public uint AddAmount(ushort materialId, uint amount)
+        {
+            if (amount == 0)
+                throw new ArgumentException("amount must be more than 0");
+
+            TradeskillMaterialEntry entry = GameTableManager.Instance.TradeskillMaterial.GetEntry(materialId);
+            if (entry == null || entry.Item2IdStatRevolution == 0)
+                throw new ArgumentNullException("entry");
```
Comment (Rawaho) [inline Source/NexusForever.WorldServer/Game/Entity/SupplySatchelManager.cs]: As above.

File: Source/NexusForever.WorldServer/Game/Entity/SupplySatchelManager.cs [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+            SendSupplySatchelStackUpdate(materialId);
+        }
+
+        /// <summary>
+        /// Moves the given amount of the given material to the player's <see cref="Inventory"/>
+        /// </summary>
+        public void MoveToInventory(ushort materialId, uint amount)
+        {
+            if (amount > tradeskillMaterials[materialId].Amount)
+                throw new ArgumentOutOfRangeException("amount");
```
Comment (Rawaho) [inline Source/NexusForever.WorldServer/Game/Entity/SupplySatchelManager.cs]: ArgumentOutOfRangeException(nameof(amount))

File: Source/NexusForever.WorldServer/Game/Entity/SupplySatchelManager.cs [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+        }
+
+        /// <summary>
+        /// Returns whether the <see cref="TradeskillMaterial"/> associated with the given <see cref="Item"/> is currently at its maximum amount
+        /// </summary>
+        public bool IsFull(Item item)
+        {
+            TradeskillMaterialEntry entry = GameTableManager.Instance.TradeskillMaterial.Entries.SingleOrDefault(i => i.Item2IdStatRevolution == item.Entry.Id);
+            if (entry == null)
+                throw new ArgumentNullException("entry");
```
Comment (Rawaho) [inline Source/NexusForever.WorldServer/Game/Entity/SupplySatchelManager.cs]: As above.

File: Source/NexusForever.WorldServer/Game/Entity/TradeskillMaterial.cs [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+            Amount = 0;
+
+            Entry = GameTableManager.Instance.TradeskillMaterial.GetEntry(MaterialId);
+
+            saveMask = TradeskillMaterialSaveMask.Create;
+        }
+
+        public void Save(CharacterContext context)
+        {
+            if (saveMask != TradeskillMaterialSaveMask.None)
```
Comment (Rawaho) [inline Source/NexusForever.WorldServer/Game/Entity/TradeskillMaterial.cs]: Try and reduce indentation.

File: Source/NexusForever.WorldServer/Network/Message/Handler/ItemHandler.cs [hunk 7]
```diff
@@ -103,5 +103,21 @@ public static void HandleItemUseDecor(WorldSession session, ClientItemUseDecor i
                 }
             }));
         }
+
+        [MessageHandler(GameMessageOpcode.ClientItemMoveToSupplySatchel)]
+        public static void HandleClient012A(WorldSession session, ClientItemMoveToSupplySatchel moveToSupplySatchel)
```
Comment (Rawaho) [inline Source/NexusForever.WorldServer/Network/Message/Handler/ItemHandler.cs]: Rename the handlers to the packet names.

File: Source/NexusForever.WorldServer/Network/Message/Handler/ItemHandler.cs [hunk 8]
```diff
@@ -103,5 +103,21 @@ public static void HandleItemUseDecor(WorldSession session, ClientItemUseDecor i
...
+        {
+            Item item = session.Player.Inventory.GetItem(moveToSupplySatchel.ItemGuid);
+            if (item == null)
+                throw new InvalidPacketValueException();
+
+            session.Player.Inventory.ItemMoveToSupplySatchel(item, moveToSupplySatchel.Amount);
+        }
+
+        [MessageHandler(GameMessageOpcode.ClientItemMoveFromSupplySatchel)]
+        public static void HandleClient0184(WorldSession session, ClientItemMoveFromSupplySatchel request)
```
Comment (Rawaho) [inline Source/NexusForever.WorldServer/Network/Message/Handler/ItemHandler.cs]: Rename the handlers to the packet names.

**Review by Rawaho — CHANGES_REQUESTED** [review]

**Comment by kirmmin** [comment]
Ready for re-review. Thanks for the pointers.

**Review by Rawaho — APPROVED** [review]

**Pull request merged**